### PR TITLE
Move BOM generation to a util within py-allspice itself

### PR DESCRIPTION
**Pull request by shrik450**
After the discussion on #31, we decided to switch from having BOM generation be an example script to a util function within py-allspice. The reasons we want to do this are:

1. We can support the BOM generation utility by versioning it as we do with py-allspice generally, so users who need use a specific version can continue to do so using standard python tooling. With an example script, any change in the logic of the script could lead to messy upgrades, especially given that the AllSpice generated JSON can theoretically change in the future.
2. We can test BOM generation using pytest as we do for the rest of py-allspice without it being questionably placed in the `examples/` directory.

Besides these, this approach ended up having the following advantages:

1. The API is defined in terms of python functions and classes, so users are not tied to the command line for something that could be invoked as part of a larger script. For those who do want to use it on the command line, we have an example wrapper script that functions  as before.
2. We are no longer bound to using JSON for the attributes mapping import. The attributes mapping is a simple data class, and users can generate it however they like - by hardcoding it, fetching it from json, yaml, toml etc.
3. COGS computation is no longer tied to py-allspice. It wasn't even in #31, but now it's even clearer that you can compute COGS with any BOM.

Two Notes:

1. The integration test for BOM generation runs fairly long, but should finish in <1m on average.
2. We're still keeping an *example* COGS computation script in the examples directory, but if we feel like that could lead to expectations of support or other issues I can remove that as part of this PR.

### CODE REVIEW

File: examples/generate_bom/generate_bom.py [hunk 1]
```diff
@@ -137,87 +66,48 @@ def extract_attributes_from_schdoc(schdoc_file_content, attributes_to_extract):
...
+    bom_rows = [
+        [
+            bom_row.description,
+            ", ".join(bom_row.designators),
+            bom_row.quantity,
+            bom_row.manufacturer,
+            bom_row.part_number,
+        ]
+        for bom_row in bom_rows
+    ]
```
Comment (danielallspice) [inline examples/generate_bom/generate_bom.py:93]: What does the for loop at the end of the assignment do?
```
 bom_rows = [
        [
            bom_row.description,
            ", ".join(bom_row.designators),
            bom_row.quantity,
            bom_row.manufacturer,
            bom_row.part_number,
        ]
        for bom_row in bom_rows
    ]
```

Comment (McRaeAlex) [reply examples/generate_bom/generate_bom.py:93]: This is a Python list comprehension https://python.land/deep-dives/list-comprehension

It's a fancy way to create a list from some computation on another list.

**Review by danielallspice — APPROVED** [review]

**Review by McRaeAlex — APPROVED** [review]
My understanding is that because we are including bom_generation in the library itself customers can expect the output of that to be stable.

In comparison the COGS script is simply an example and we can choose to break it or modify its outputs at any time correct?

**Comment by shrik450** [comment]
Yes, but maybe we need a better way to communicate that. I think in the discussion we concluded that the best way to incorporate these would be to have plug and play CI actions that we version, so if we make one of those for COGS later we can version and support it accordingly.

File: allspice/utils/bom_generation.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+@dataclass
+class SchdocComponent:
+    description: str
+    manufacturer: str
+    part_number: str
+    designator: str
+
+
+@dataclass
+class PcbComponent:
```
Comment (jtran) [inline allspice/utils/bom_generation.py:23]: Should this be `PcbdocComponent` to be consistent with `SchdocComponent`? Is this specific to the file type?

Comment (shrik450) [reply allspice/utils/bom_generation.py:23]: No, it isn't specific to the file type. Thanks for catching it, I'll rename `SchdocComponent` to `SchematicComponent`.

**Review by jtran — APPROVED** [review]
Nice.

**Review by McRaeAlex — APPROVED** [review]

**Pull request merged**